Merge latest upstream changes

vocksel opened the pull request:
This PR merges the latest changes from commit 0d5efe33bc151377cbca86f6513190c6f4f6af29 in Roblox/jest-roblox

A new package JestMockGenv is included with these changes, and I'm hitting some build issues because of it. Seems like this package may need to be published to the jsdotlua npm scope first.

Checklist before submitting:
* [ ] Tidy up merge artifacts
* [ ] Added/updated relevant tests
* [ ] Added/updated documentation

Commit:
APT-328 - Implement jest.spyOn (#382)

## Problem
[APT-328](https://roblox.atlassian.net/browse/APT-328)

## Solution
Align and complete implementation of `spyOn` function

## Testing
Aligned a number of tests, including further tests for already-ported
logic.

Commit:
#nojira Fix misconception in comment (#383)

Commit:
UIBLOX-667: Update jest mock type (#385)

The type in the export is incorrect

See here

https://github.com/Roblox/jest-roblox-internal/blob/master/src/jest-runtime/src/init.lua#L2450

Updating and bumping to be moduleScript instead of string

---------

Commit:
APT-1452 dom to fs path resolution in jest-runner for jest-vscode reconciler (#373)

Commit:
Release 3.7.0 (#386)

Updating changelog and bumping release version for 3.7.0. Needed for the vscode extension.

Commit:
Mock task wait (#388)

Mocking task.wait directly in jest as per suggestion of @LordOfSpelunky

Commit:
Configure Jest to use the Registry (#390)

## Problem
Projects using the registry cannot consume new versions of Jest

## Solution
Configre Jest to use the registry

## Testing
Describe how the solution was tested (ideally unit tests). If done
manually, describe the process and outcome.

## Checklist (remove if not applicable)
- [ ] For new modules, has a README.md file with notes about the
translation
- [ ] All deviations are notated in code with `-- ROBLOX deviation:
comment`
- [ ] Translated files include a comment with a link to the upstream
file

Commit:
#nojira fix mismatched test paths between reporter and runner (#395)

Commit:
v3.8.1 (#396)

Commit:
APT-597 mock lua globals (#397)

## Problem
The `spyOn` function does not currently work with Lua globals because
those tables are not mockable in the usual way.

## Solution
Adds `jest.globalEnv` which can be used with `spyOn` to mock Lua globals
in a test environment. Allows access to the unmocked globals as members
too.

## Testing
Added new unit test cases to cover API surface.

## Checklist (remove if not applicable)
- [x] For new modules, has a README.md file with notes about the
translation
- [x] All deviations are notated in code with `-- ROBLOX deviation:
comment`
- [x] Translated files include a comment with a link to the upstream
file

Commit:
APT-1775 Optimistic protected instance property indexing (#398)

## Problem
Jest assumes that all properties listed in the Roblox API are safe to
index. This is not true when encountering properties that are only
accessible to privileged code while Jest is running in a less privileged
context, or when otherwise dealing with properties that have the
potential to error on index.

## Solution
Hide details of the Roblox API inside of RobloxInstance, and instead
only expose 'safe' APIs that protect around all instance reads and
writes, and expose only the properties which are accessible with the
script's current capabilities.

## Testing
Unit tests and snapshots.

Commit:
APT-1023 APT-1778 Update links and fix static assets (#400)

Moved all jest docs links to point to the 27.x archived version of the
jest docs and updated static assets correctly the docusaurus way.

Commit:
APT-1808 update manifests, release 3.9.0 (#402)

Commit:
v3.9.1, fix type in runtime (#403)

Commit:
APT-1797 More helpful error when requiring `JestGlobals` outside a test environment (#405)

## Problem
When Jest 2 users are in the midst of updating to Jest 3, they might
attempt to pull in Jest 3 globals.

https://roblox.slack.com/archives/C04N86WJA4R/p1721837297468429?thread_ts=1721815127.913289&cid=C04N86WJA4R

## Solution
Update the JestGlobals error to include a tip about mixing Jest 2 and
Jest 3.

## Testing
Unit tests.

Commit:
APT-1523 fallback to loadstring if loadmodule not enabled (#392)

vocksel committed:
Merge remote-tracking branch 'roblox/master' into upgrade

vocksel committed:
Setup JestMockGenv for publish

vocksel committed:
Fix require paths

Commit:
APT-1766 Better error snapshots (#401)

## Problem
When an error with a stack trace is saved to a snapshot, superficial or
unrelated changes in the codebase can cause the snapshot test to fail,
because line numbers or file names in the stack trace can change. In
particular, this occurs for some of Jest's own snapshot tests.

## Solution
Introduce a new `redactStackTrace` function to `RobloxShared` which
replaces stack traces with fixed-size dummy contents, similar to how
`pruneDeps` cleans up stack trace lines from stack traces today. The
final result is intended to be formatted exactly like a valid stack
trace, except it is guaranteed to be completely decoupled from the
actual stack trace it replaces, allowing for the stack trace to be
snapshotted safely.

Use `redactStackTrace` in the flaky snapshot tests in
`errorParsing.spec` which were causing problems before.

## Testing
Unit tests and snapshots.

Commit:
APT-1805 add JestBenchmark to the documentation page (#406)

Commit:
APT-1837 Protect against bad plugin table reads (#407)

## Problem
When adding `pretty-format` plugins, it's possible to accidentally miss
adding the plugin to the `prettyFormat.plugins` table. However, other
code might still depend on the plugin being accessible from that table,
which will silently fail.

## Solution
Add an `__index` metatable which throws an error, so that when a plugin
can't be found in the table, the issue is escalated right away.

## Testing
Unit test.

## Checklist
- [x] All deviations are notated in code with `-- ROBLOX deviation:
comment`

Commit:
APT-1826 Stabilise RobloxInstance "serialises Folder" test. (#408)

## Problem
RobloxInstance "serialises Folder" test should not fail when source code
in RobloxInstance changes.

## Solution
Copy a different folder of code into the tests directory as an
unchanging target. Then, point the test at that.

## Testing
Snapshot tests test themselves 🤠

Commit:
Release 3.10.0 (#411)

jeparlefrancais commented on src/jest-mock-genv/package.json:
```diff
@@ -0,0 +1,21 @@
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/jsdotlua/jest-lua.git",
+    "directory": "src/jest-mock"
```
```suggestion
    "directory": "src/jest-mock-genv"
```

jeparlefrancais reviewed (approved):
Minor thing! Other than this it looks good to me 👍 

Thanks a lot this is very valuable work 🎉

vocksel committed:
Point to right directory for jest-mock-genv

Co-authored-by: jeparlefrancais <35781636+jeparlefrancais@users.noreply.github.com>

vocksel committed:
Run yarn install

Commit:
bump rbx-aged-tool (#412)

https://github.com/Roblox/rbx-aged-tool/releases/tag/5.7.10

Commit:
APT-1913 Inject alike types (#409)

## Problem
Right now, `spyOn` injects a table object when spying on objects. This
table has a different type and overall behaves quite differently from
the function it's meant to replace, and there's only so much we can do
to disguise it in Luau.

## Solution
Move away from injecting 'mock objects'. Instead, inject 'alike types'
wherever possible - in this case, inject a function when spying on a
function.

## Testing
Unit & snapshot tests.

## Checklist (remove if not applicable)
- [x] For new modules, has a README.md file with notes about the
translation
- [x] All deviations are notated in code with `-- ROBLOX deviation:
comment`
- [x] Translated files include a comment with a link to the upstream
file

vocksel committed:
Update package dependencies to include jest-mock-genv

vocksel committed:
Add linker module to roblox-model

vocksel committed:
Merge remote-tracking branch 'roblox/master' into upgrade

vocksel committed:
Adding back newlines

vocksel committed:
Spruce up lockfile

vocksel committed:
Fix require paths

Commit:
APT-2003 Disallow require by string for now (#414)

## Problem
In future editions of Luau, require-by-string will be enabled, allowing
users to pass in string paths to `require()` which are resolved
internally to a ModuleScript instance.

## Solution
Disable this feature in Jest, because we do not yet have functionality
to match this. Throws a useful error instead.

## Testing
Unit tests.

## Checklist
- [x] All deviations are notated in code with `-- ROBLOX deviation:
comment`

jeparlefrancais committed:
Fix a few remaining Roblox requires

jeparlefrancais committed:
Bump darklua version

jeparlefrancais committed:
Specify node version 22

Latest recent node version breaks the node setup and yarn won't install
things. Switching to version 22 solves this.

Commit:
APT-1914 spyOn callable tables (#413)

## Problem
Callable tables are not compatible with the `spyOn` method.

## Solution
Allow callable tables to be injected by `spyOn`, cloning their data and
metatable while replacing the `__call` with a forwarding function to the
mock object.

## Testing
Unit tests and snapshots

## Checklist (remove if not applicable)
- [x] For new modules, has a README.md file with notes about the
translation
- [x] All deviations are notated in code with `-- ROBLOX deviation:
comment`

Commit:
APT-2175 Quantum stability for RobloxInstance (#416)

## Problem
Sometimes, `RobloxInstance.listProps` is prone to returning incorrect
property listings because of quantum UI bugs.

## Solution
Perform a "cold read" of all properties to allow any quantum UI
calculations to run, then do a "warm read" to collect all the properties
once they've settled into their proper values.

This behaviour is optional. Existing features do not utilise this "warm
read" behaviour for backwards compat.

## Testing
Unit tests and snapshots

Commit:
APT-2004 Add instance mocking classes (#418)

## Problem

Jest currently lacks the facilities to adequately mock and spy on the
behaviour of `Instance` objects, because Jest's existing methods of
spying on objects are invasive.

Additionally, it's unclear how our existing mocking tech could expand to
cover the nuances of instances well.

## Solution

Introduce new internal classes for creating and working with mock
instances:

- `InstanceProxy` implementing basic proxying and mocking support
- `DataModelMocker` to house complex mocking logic for instance <->
instance operations

The implementation of `DataModelMocker` will remain pretty barebones for
now, just acting as a simple cache for proxies, because not much
functionality is needed here for the MVP of `GetService` mocks. In the
future, if we implement a translation layer to properly sandbox
instances, this will likely need to expand to support more complex data
model operations, e.g. returning already-mocked instances when indexing
into children or `.Parent`, so that `game.Workspace.Parent == game`, for
example.

## Testing
Unit tests.

vocksel committed:
Merge remote-tracking branch 'upstream/master' into upgrade

vocksel committed:
Add jest-mock-rbx package

vocksel committed:
Update release workflow

vocksel committed:
Fix build

vocksel committed:
Get tests running via `yarn test`

MayGo commented:
Would really love if this get merged. Want to run it in Open Cloud Engine: https://devforum.roblox.com/t/beta-open-cloud-engine-api-for-executing-luau/3172185/67

grilme99 commented:
I'm going to go ahead and merge this despite the failing tests. I combed through and the test failures are due to the change in file structure in our fork, and aren't actually broken functionality. I don't have the time to fix the tests and snapshots, but need to get this merged in ASAP.
